src/v/cloud_topics/tests/end_to_end_test.cc

```diff
@@ -119,3 +122,89 @@ TEST_F(e2e_fixture, test_l0_path) {
         }
     }
 }
+
+TEST_F(e2e_fixture, timequery) {
+    lconf().log_message_timestamp_after_max_ms.set_value(
+      serde::max_serializable_ms);
+    auto unset_cluster_config = ss::defer([&] {
+        lconf().log_message_timestamp_after_max_ms.set_value(
+          lconf().log_message_timestamp_after_max_ms.default_value());
+    });
+
+    // Units we use in this test for relative timestamps is *hours* as to make
+    // sure the slow uploading into cloud topics doesn't effect the results.
+    constexpr int64_t hour_in_milli = 60L * 60 * 1000;
+    struct timequery_batch_spec {
+        std::vector<int64_t> relative_timestamps;
+        bool broker_time = false; // append time or create time?
+    };
+
+    std::vector<timequery_batch_spec> data_spec{
+      {.relative_timestamps = {-60, -50, -40, -30}},
+      {.relative_timestamps = {-25, -24, -25, -25}},
+      {.relative_timestamps = {-29, -28, -27}},
+      {.relative_timestamps = {-19, -20, -19, -20, -19}},
+      // For broker time, we should be ignoring the timestamp deltas
+      {.relative_timestamps = {0, 1, 2, 3, 5}, .broker_time = true},
+      {.relative_timestamps = {16, 17, 17, 18, 16, 20}},
+    };
+    model::timestamp now = model::timestamp::now();
+    auto* producer = make_producer();
+    for (const auto& spec : data_spec) {
+        model::batch_builder builder;
+        for (auto rel_ts : spec.relative_timestamps) {
+            builder.add_record(
+              model::record(/*attributes=*/{},
+                            /*timestamp_delta=*/rel_ts * hour_in_milli,
+                            /*offset_delta=*/builder.num_records(),
+                            /*key=*/std::nullopt,
+                            /*value=*/std::nullopt,
+                            /*hdrs=*/{}));
+        }
+        auto timestamp_type = spec.broker_time
+                                ? model::timestamp_type::append_time
+                                : model::timestamp_type::create_time;
+        builder.set_batch_timestamp(timestamp_type, now);
+        // Set the timestamp type for the topic (well cluster) so that we can
+        // toggle between the different types of append vs create time.
+        lconf().log_message_timestamp_type.set_value(timestamp_type);
+        auto unset_cluster_config = ss::defer([&] {
+            lconf().log_message_timestamp_type.set_value(
+              lconf().log_message_timestamp_type.default_value());
+        });
+        producer->produce_to_partition(ntp, builder.build_sync()).get();
+    }
```

Comment:
Would be great if we could force reconcile to L1 too

Reply:
Agreed. I will think about how we can do that.